`-Ysafe-init-global` not working with Scala 2 library TASTy

## Problematic tests

* `tests/init-global/neg/patmat.scala`
* `tests/init-global/neg/t9312.scala`
* `tests/init-global/neg/unapplySeq-implicit-arg.scala`
* `tests/init-global/neg/unapplySeq-implicit-arg2.scala`
* `tests/init-global/neg/unapplySeq-implicit-arg3.scala`

## Reproduction 1


```
sbt> set ThisBuild/Build.useScala2LibraryTasty := true
sbt> scala3-bootstrapped/testCompilation tests/init-global/neg/t9312.scala
```

## Reproduction 2

<details><summary>Without library TASTy</summary>

```
sbt> set ThisBuild/Build.useScala2LibraryTasty := false
sbt> scala3-bootstrapped/scalac -Ysafe-init-global tests/init-global/neg/t9312.scala
```

```scala
-- Warning: tests/init-global/neg/t9312.scala:18:11 ----------------------------
18 |    object Child1 extends Child     // error
   |           ^
   |Cyclic initialization: object Child1 -> object Parent -> object Child1. Calling trace:
   |-> object Child1 extends Child     // error [ t9312.scala:18 ]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |-> trait Child {    [ t9312.scala:12 ]
   |   ^
   |-> val parent = Parent      [ t9312.scala:14 ]
   |                ^^^^^^
   |-> object Parent {                    // error      [ t9312.scala:11 ]
   |   ^
   |-> final val children = Set(Child1, Child2) [ t9312.scala:21 ]
   |                            ^^^^^^
-- Warning: tests/init-global/neg/t9312.scala:11:9 -----------------------------
11 |  object Parent {                    // error
   |         ^
   |Cyclic initialization: object Parent -> object Child2 -> object Parent. Calling trace:
   |-> object Parent {                    // error      [ t9312.scala:11 ]
   |   ^
   |-> final val children = Set(Child1, Child2) [ t9312.scala:21 ]
   |                                    ^^^^^^
   |-> object Child2 extends Child      [ t9312.scala:19 ]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |-> trait Child {    [ t9312.scala:12 ]
   |   ^
   |-> val parent = Parent      [ t9312.scala:14 ]
   |                ^^^^^^
2 warnings found
```
</details> 

<details><summary>With library TASTy</summary>

```
sbt> set ThisBuild/Build.useScala2LibraryTasty := true
sbt> scala3-bootstrapped/scalac -Ysafe-init-global tests/init-global/neg/t9312.scala
```

```scala
-- Warning: tests/init-global/neg/t9312.scala:18:11 ----------------------------
18 |    object Child1 extends Child     // error
   |           ^
   |Cyclic initialization: object Child1 -> object Parent -> object Child1. Calling trace:
   |-> object Child1 extends Child     // error [ t9312.scala:18 ]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |-> trait Child {    [ t9312.scala:12 ]
   |   ^
   |-> val parent = Parent      [ t9312.scala:14 ]
   |                ^^^^^^
   |-> object Parent {                    // error      [ t9312.scala:11 ]
   |   ^
   |-> final val children = Set(Child1, Child2) [ t9312.scala:21 ]
   |                            ^^^^^^
-- Warning: tests/init-global/neg/t9312.scala:11:9 -----------------------------
11 |  object Parent {                    // error
   |         ^
   |Cyclic initialization: object Parent -> object Child2 -> object Parent. Calling trace:
   |-> object Parent {                    // error      [ t9312.scala:11 ]
   |   ^
   |-> final val children = Set(Child1, Child2) [ t9312.scala:21 ]
   |                                    ^^^^^^
   |-> object Child2 extends Child      [ t9312.scala:19 ]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |-> trait Child {    [ t9312.scala:12 ]
   |   ^
   |-> val parent = Parent      [ t9312.scala:14 ]
   |                ^^^^^^
Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating a field before its initialization: variable mutationCount. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.mutationCount = ArrayBuffer.this.mutationCount.+(1)

Mutating a field before its initialization: variable size0. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.size0 = ArrayBuffer.this.length.+(elemsLength)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating a field before its initialization: variable mutationCount. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.mutationCount = ArrayBuffer.this.mutationCount.+(1)

Mutating a field before its initialization: variable size0. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.size0 = ArrayBuffer.this.length.+(elemsLength)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating a field before its initialization: variable mutationCount. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.mutationCount = ArrayBuffer.this.mutationCount.+(1)

Mutating a field before its initialization: variable size0. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.size0 = ArrayBuffer.this.length.+(elemsLength)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating a field before its initialization: variable mutationCount. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.mutationCount = ArrayBuffer.this.mutationCount.+(1)

Mutating a field before its initialization: variable size0. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.size0 = ArrayBuffer.this.length.+(elemsLength)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.clone()

Reading mutable state of object Array during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> BitmapIndexedSetNode.this.originalHashes.apply(index)

Reading mutable state of object ArrayBuffer during initialization of object Parent.
Reading mutable state of other static objects is forbidden as it breaks initialization-time irrelevance. Calling trace: 
-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.apply(n)

Mutating object ArrayBuffer during initialization of object Parent.
Mutating other static objects during the initialization of one static object is forbidden. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> ArrayBuffer.this.array.update(index, elem.asInstanceOf[AnyRef])

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.appended[A](element)

Mutating a field before its initialization: variable content. Calling trace:

-> object Parent {                    // error  [ t9312.scala:11 ]
   ^
-> final val children = Set(Child1, Child2)     [ t9312.scala:21 ]
                        ^^^^^^^^^^^^^^^^^^^
-> hc.content = hc.content.updated[A](index, element)

86 warnings found
```
</details> 


## Question

What trigers the extra warnings? It is possible that the TASTy has some inconsitent information. It is also possible that the TASTy has more information, for example, now we know which classes have a pure constuctor.

Tried the latest master with

``` Scala
sbt:scala3> set ThisBuild/Build.useScala2LibraryTasty := true
sbt:scala3> scala3-compiler-bootstrapped/scalac -Ysafe-init-global -d out tests/init-global/neg/t9312.scala
```

And get a crash in loading the TASTy:

``` Scala
TASTy file scala/LowPriorityImplicits.tasty is broken, reading aborted with class java.lang.ArrayIndexOutOfBoundsException
  Index 6583 out of bounds for length 6583
Exception in thread "main" java.io.IOException: TASTy file scala/$less$colon$less.tasty is broken, reading aborted with class java.lang.ArrayIndexOutOfBoundsException
  Index 9345 out of bounds for length 9345
	at dotty.tools.dotc.core.TastyLoader.handleUnpicklingExceptions(SymbolLoaders.scala:448)
	at dotty.tools.dotc.core.TastyLoader.<init>(SymbolLoaders.scala:423)
	at dotty.tools.dotc.config.JavaPlatform.newTastyLoader(JavaPlatform.scala:71)
	at dotty.tools.dotc.core.SymbolLoaders$.initializeFromClassPath(SymbolLoaders.scala:201)
	at dotty.tools.dotc.core.SymbolLoaders$PackageLoader.enterClasses$$anonfun$1(SymbolLoaders.scala:276)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.VectorStatics$.foreachRec(Vector.scala:2124)
	at scala.collection.immutable.Vector.foreach(Vector.scala:2130)
	at dotty.tools.dotc.core.SymbolLoaders$PackageLoader.enterClasses(SymbolLoaders.scala:276)
	at dotty.tools.dotc.core.SymbolLoaders$PackageLoader.doComplete$$anonfun$1$$anonfun$1(SymbolLoaders.scala:295)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at dotty.tools.dotc.core.SymbolLoaders$PackageLoader$PackageScope.lookupEntry(SymbolLoaders.scala:236)
	at dotty.tools.dotc.core.SymDenotations$PackageClassDenotation.proceedWithEnter(SymDenotations.scala:2575)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.enter(SymDenotations.scala:2049)
	at dotty.tools.dotc.core.Symbols$Symbol.entered(Symbols.scala:210)
	at dotty.tools.dotc.core.Definitions.enterSpecialPolyClass(Definitions.scala:88)
	at dotty.tools.dotc.core.Definitions.RepeatedParamClass(Definitions.scala:640)
	at dotty.tools.dotc.core.Definitions.syntheticScalaClasses(Definitions.scala:2133)
	at dotty.tools.dotc.core.Definitions.syntheticCoreClasses(Definitions.scala:2141)
	at dotty.tools.dotc.core.Definitions.init(Definitions.scala:2157)
	at dotty.tools.dotc.core.Contexts$ContextBase.initialize(Contexts.scala:902)
	at dotty.tools.dotc.core.Contexts$Context.initialize(Contexts.scala:523)
	at dotty.tools.dotc.Run.rootContext(Run.scala:450)
	at dotty.tools.dotc.Run.<init>(Run.scala:471)
	at dotty.tools.dotc.Compiler.newRun(Compiler.scala:172)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
	at dotty.tools.dotc.Driver.process(Driver.scala:197)
	at dotty.tools.dotc.Driver.process(Driver.scala:165)
	at dotty.tools.dotc.Driver.process(Driver.scala:177)
	at dotty.tools.dotc.Driver.main(Driver.scala:207)
	at dotty.tools.dotc.Main.main(Main.scala)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 9345 out of bounds for length 9345
	at dotty.tools.tasty.TastyReader.readLongNat(TastyReader.scala:78)
	at dotty.tools.tasty.TastyReader.readNat(TastyReader.scala:64)
	at dotty.tools.tasty.TastyReader.readNameRef(TastyReader.scala:118)
	at dotty.tools.dotc.core.tasty.AttributeUnpickler.attributes$lzyINIT1(AttributeUnpickler.scala:24)
	at dotty.tools.dotc.core.tasty.AttributeUnpickler.attributes(AttributeUnpickler.scala:14)
	at dotty.tools.dotc.core.tasty.DottyUnpickler.$init$$$anonfun$1(DottyUnpickler.scala:61)
	at scala.Option.map(Option.scala:242)
	at dotty.tools.dotc.core.tasty.DottyUnpickler.<init>(DottyUnpickler.scala:61)
	at dotty.tools.dotc.core.TastyLoader.$init$$$anonfun$1(SymbolLoaders.scala:423)
	at dotty.tools.dotc.core.TastyLoader.handleUnpicklingExceptions(SymbolLoaders.scala:439)
	... 32 more

```

It didn't happen in the previous reproduction of issue. Not sure if something has changed in between. Any ideas? /cc: @nicolasstucki

--- a/compiler/src/dotty/tools/dotc/transform/init/Objects.scala
+++ b/compiler/src/dotty/tools/dotc/transform/init/Objects.scala
@@ -875,7 +875,7 @@ object Objects:
    * @param ctor        The symbol of the target constructor.
    * @param args        The arguments passsed to the constructor.
    */
-  def instantiate(outer: Value, klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo], inKlass: ClassSymbol): Contextual[Value] = log("instantiating " + klass.show + ", outer = " + outer + ", args = " + args.map(_.value.show), printer, (_: Value).show) {
+  def instantiate(outer: Value, klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo]): Contextual[Value] = log("instantiating " + klass.show + ", outer = " + outer + ", args = " + args.map(_.value.show), printer, (_: Value).show) {
     outer match
 
     case _ : Fun | _: OfArray  =>
@@ -914,7 +914,7 @@ object Objects:
         instance
 
     case ValueSet(values) =>
-      values.map(ref => instantiate(ref, klass, ctor, args, inKlass)).join
+      values.map(ref => instantiate(ref, klass, ctor, args)).join
   }
 
   /** Handle local variable definition, `val x = e` or `var x = e`.
@@ -1088,7 +1088,7 @@ object Objects:
         val cls = tref.classSymbol.asClass
         withTrace(trace2) {
           val outer = outerValue(tref, thisV, klass)
-          instantiate(outer, cls, ctor, args, klass)
+          instantiate(outer, cls, ctor, args)
         }
 
       case Apply(ref, arg :: Nil) if ref.symbol == defn.InitRegionMethod =>

--- a/compiler/src/dotty/tools/dotc/transform/init/Objects.scala
+++ b/compiler/src/dotty/tools/dotc/transform/init/Objects.scala
@@ -565,7 +565,7 @@ object Objects:
 
   // --------------------------- domain operations -----------------------------
 
-  type ArgInfo = TraceValue[Value]
+  case class ArgInfo(value: Value, trace: Trace, tree: Tree)
 
   extension (a: Value)
     def join(b: Value): Value =
@@ -875,7 +875,7 @@ object Objects:
    * @param ctor        The symbol of the target constructor.
    * @param args        The arguments passsed to the constructor.
    */
-  def instantiate(outer: Value, klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo]): Contextual[Value] = log("instantiating " + klass.show + ", outer = " + outer + ", args = " + args.map(_.value.show), printer, (_: Value).show) {
+  def instantiate(outer: Value, klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo], inKlass: ClassSymbol): Contextual[Value] = log("instantiating " + klass.show + ", outer = " + outer + ", args = " + args.map(_.value.show), printer, (_: Value).show) {
     outer match
 
     case _ : Fun | _: OfArray  =>
@@ -884,9 +884,14 @@ object Objects:
 
     case outer: (Ref | Cold.type | Bottom.type) =>
       if klass == defn.ArrayClass then
-        val arr = OfArray(State.currentObject, summon[Regions.Data])
-        Heap.writeJoin(arr.addr, Bottom)
-        arr
+        args.head.tree.tpe match
+          case ConstantType(Constants.Constant(0)) =>
+            // new Array(0)
+            Bottom
+          case _ =>
+            val arr = OfArray(State.currentObject, summon[Regions.Data])
+            Heap.writeJoin(arr.addr, Bottom)
+            arr
       else
         // Widen the outer to finitize the domain. Arguments already widened in `evalArgs`.
         val (outerWidened, envWidened) =
@@ -909,7 +914,7 @@ object Objects:
         instance
 
     case ValueSet(values) =>
-      values.map(ref => instantiate(ref, klass, ctor, args)).join
+      values.map(ref => instantiate(ref, klass, ctor, args, inKlass)).join
   }
 
   /** Handle local variable definition, `val x = e` or `var x = e`.
@@ -1083,7 +1088,7 @@ object Objects:
         val cls = tref.classSymbol.asClass
         withTrace(trace2) {
           val outer = outerValue(tref, thisV, klass)
-          instantiate(outer, cls, ctor, args)
+          instantiate(outer, cls, ctor, args, klass)
         }
 
       case Apply(ref, arg :: Nil) if ref.symbol == defn.InitRegionMethod =>
@@ -1328,7 +1333,7 @@ object Objects:
           case _ => List()
 
         val implicitArgsAfterScrutinee = evalArgs(implicits.map(Arg.apply), thisV, klass)
-        val args = implicitArgsBeforeScrutinee(fun) ++ (TraceValue(scrutinee, summon[Trace]) :: implicitArgsAfterScrutinee)
+        val args = implicitArgsBeforeScrutinee(fun) ++ (ArgInfo(scrutinee, summon[Trace], EmptyTree) :: implicitArgsAfterScrutinee)
         val unapplyRes = call(receiver, funRef.symbol, args, funRef.prefix, superType = NoType, needResolve = true)
 
         if fun.symbol.name == nme.unapplySeq then
@@ -1425,7 +1430,7 @@ object Objects:
       // call .lengthCompare or .length
       val lengthCompareDenot = getMemberMethod(scrutineeType, nme.lengthCompare, lengthCompareType)
       if lengthCompareDenot.exists then
-        call(scrutinee, lengthCompareDenot.symbol, TraceValue(Bottom, summon[Trace]) :: Nil, scrutineeType, superType = NoType, needResolve = true)
+        call(scrutinee, lengthCompareDenot.symbol, ArgInfo(Bottom, summon[Trace], EmptyTree) :: Nil, scrutineeType, superType = NoType, needResolve = true)
       else
         val lengthDenot = getMemberMethod(scrutineeType, nme.length, lengthType)
         call(scrutinee, lengthDenot.symbol, Nil, scrutineeType, superType = NoType, needResolve = true)
@@ -1433,7 +1438,7 @@ object Objects:
 
       // call .apply
       val applyDenot = getMemberMethod(scrutineeType, nme.apply, applyType(elemType))
-      val applyRes = call(scrutinee, applyDenot.symbol, TraceValue(Bottom, summon[Trace]) :: Nil, scrutineeType, superType = NoType, needResolve = true)
+      val applyRes = call(scrutinee, applyDenot.symbol, ArgInfo(Bottom, summon[Trace], EmptyTree) :: Nil, scrutineeType, superType = NoType, needResolve = true)
 
       if isWildcardStarArgList(pats) then
         if pats.size == 1 then
@@ -1444,7 +1449,7 @@ object Objects:
         else
           // call .drop
           val dropDenot = getMemberMethod(scrutineeType, nme.drop, applyType(elemType))
-          val dropRes = call(scrutinee, dropDenot.symbol, TraceValue(Bottom, summon[Trace]) :: Nil, scrutineeType, superType = NoType, needResolve = true)
+          val dropRes = call(scrutinee, dropDenot.symbol, ArgInfo(Bottom, summon[Trace], EmptyTree) :: Nil, scrutineeType, superType = NoType, needResolve = true)
           for pat <- pats.init do evalPattern(applyRes, pat)
           evalPattern(dropRes, pats.last)
         end if
@@ -1546,7 +1551,7 @@ object Objects:
         case _ =>
           res.widen(1)
 
-      argInfos += TraceValue(widened, trace.add(arg.tree))
+      argInfos += ArgInfo(widened, trace.add(arg.tree), arg.tree)
     }
     argInfos.toList
 
@@ -1644,7 +1649,7 @@ object Objects:
             // The parameter check of traits comes late in the mixin phase.
             // To avoid crash we supply hot values for erroneous parent calls.
             // See tests/neg/i16438.scala.
-            val args: List[ArgInfo] = ctor.info.paramInfoss.flatten.map(_ => new ArgInfo(Bottom, Trace.empty))
+            val args: List[ArgInfo] = ctor.info.paramInfoss.flatten.map(_ => new ArgInfo(Bottom, Trace.empty, EmptyTree))
             extendTrace(superParent) {
               superCall(tref, ctor, args, tasks)
             }

--- a/tests/init-global/pos/array-size-zero.scala
+++ b/tests/init-global/pos/array-size-zero.scala
@@ -1,0 +1,10 @@
+object A:
+  val emptyArray = new Array(0)
+
+object B:
+  def build(data: Int*) =
+    if data.size == 0 then A.emptyArray else Array(data)
+
+  val arr = build(5, 6)
+  val first = arr(0)
+

